l1/ecl_l1_pos_controller.cpp

```diff
@@ -200,107 +222,95 @@ ECL_L1_Pos_Controller::navigate_waypoints(const Vector2f &vector_A, const Vector
+	_target_bearing = get_bearing_to_next_waypoint((double)vector_curr_position(0), (double)vector_curr_position(1),
+			  (double)vector_A(0), (double)vector_A(1));
 
-	/* ground speed, enforce minimum of 0.1 m/s to avoid singularities */
-	float ground_speed = math::max(ground_speed_vector.length(), 0.1f);
+	/* enforce a minimum ground speed/airspeed of 0.1 m/s to avoid singularities */
+	const float ground_speed = math::max(ground_speed_vector.length(), 0.1f);
+	Vector2f airspeed_vector = ground_speed_vector - wind_speed_vector;
+	const float airspeed = math::max(airspeed_vector.length(), 0.1f);
 
 	/* calculate the L1 length required for the desired period */
 	_L1_distance = _L1_ratio * ground_speed;
 
 	/* calculate the vector from waypoint A to current position */
 	Vector2f vector_A_to_airplane = get_local_planar_vector(vector_A, vector_curr_position);
 
-	Vector2f vector_A_to_airplane_unit;
-
-	/* prevent NaN when normalizing */
-	if (vector_A_to_airplane.length() > FLT_EPSILON) {
-		/* store the normalized vector from waypoint A to current position */
-		vector_A_to_airplane_unit = vector_A_to_airplane.normalized();
-
-	} else {
-		vector_A_to_airplane_unit = vector_A_to_airplane;
+	/* avoid erroneous bearing calculations when aircraft is in center of loiter
+	 * -- set arbitrary direction until out of center */
+	if (vector_A_to_airplane.length() < 0.1f) {
+		vector_A_to_airplane(0) = 0.1;
+		vector_A_to_airplane(1) = 0.0;
 	}
 
-	/* calculate eta angle towards the loiter center */
-
-	/* velocity across / orthogonal to line from waypoint to current position */
-	float xtrack_vel_center = vector_A_to_airplane_unit % ground_speed_vector;
-	/* velocity along line from waypoint to current position */
-	float ltrack_vel_center = - (ground_speed_vector * vector_A_to_airplane_unit);
-	float eta = atan2f(xtrack_vel_center, ltrack_vel_center);
-	/* limit eta to 90 degrees */
-	eta = math::constrain(eta, -M_PI_F / 2.0f, +M_PI_F / 2.0f);
+	const float norm_vector_A_to_airplane = vector_A_to_airplane.length();
 
-	/* calculate the lateral acceleration to capture the center point */
-	float lateral_accel_sp_center = _K_L1 * ground_speed * ground_speed / _L1_distance * sinf(eta);
-
-	/* for PD control: Calculate radial position and velocity errors */
-
-	/* radial velocity error */
-	float xtrack_vel_circle = -ltrack_vel_center;
 	/* radial distance from the loiter circle (not center) */
-	float xtrack_err_circle = vector_A_to_airplane.length() - radius;
+	const float xtrack_err_circle = norm_vector_A_to_airplane - radius;
+
+	/* check circle tracking feasibility */
+	if (_L1_distance > radius && fabsf(xtrack_err_circle) <= _L1_distance) {
+		// re-calculate L1 ratio & distance (effectively reduce period to allow tracking small radii)
+		// NOTE: when far from the loiter radius, operator defined period and damping are unaffected
+		_L1_ratio = math::max(fabsf(xtrack_err_circle), radius) / ground_speed;
```

Comment:
Reducing L1 distance does increase the guidance loop gain which can produce limit cycling in roll if the roll angle AP has not been tuned correctly and the vehicle has excessive lag from demanded to achieved lateral acceleration. This can be caused by lack of tuning or by adverse yaw. That is the main disadvantage to this approach and the main reason why the original method  of tracking circles was implemented.

We need to update user documentation so that they will know what to do if this happens.